editor/src/core/shared/project-contents-dependencies.ts

```diff
@@ -0,0 +1,170 @@
+import { emptySet } from './set-utils'
+
+export function getDirectReverseDependencies(
+  projectContents: ProjectContentTreeRoot,
+  nodeModules: NodeModules,
+): MapLike<Array<string>> {
+  let result: MapLike<Array<string>> = {}
+  walkContentsTree(projectContents, (fullPath, file) => {
+    if (isTextFile(file)) {
+      const parseResult = file.fileContents.parsed
+      if (isParseSuccess(parseResult)) {
+        for (const importSource of Object.keys(parseResult.imports)) {
+          const resolveResult = resolveModule(projectContents, nodeModules, fullPath, importSource)
+          switch (resolveResult.type) {
+            case 'RESOLVE_SUCCESS':
+              const successResult = resolveResult.success
+              switch (successResult.file.type) {
+                case 'ES_CODE_FILE':
+                  // Check if the file is part of the project.
+                  if (getContentsTreeFileFromString(projectContents, successResult.path) != null) {
+                    // Dependency, so record the connection.
+                    result = addToMapOfArraysUnique(result, successResult.path, fullPath)
+                  }
+                  break
+                case 'ES_REMOTE_DEPENDENCY_PLACEHOLDER':
+                  // Do nothing for these.
+                  break
+                default:
+                  const _exhaustiveCheck: never = successResult.file
+                  throw new Error(`Unhandled case ${JSON.stringify(successResult.file)}`)
+              }
+              break
+            case 'RESOLVE_NOT_PRESENT':
+              // Do nothing for now...
+              break
+            default:
+              const _exhaustiveCheck: never = resolveResult
+              throw new Error(`Unhandled case ${JSON.stringify(resolveResult)}`)
+          }
+        }
+      }
+    }
+  })
+  return result
+}
+
+export function getTransitiveReverseDependencies(
+  projectContents: ProjectContentTreeRoot,
+  nodeModules: NodeModules,
+  filenamesThatChanged: Array<string>,
+): Array<string> {
+  if (filenamesThatChanged.length > 0) {
+    let result: Set<string> = emptySet()
+    const directReverseDeps = getDirectReverseDependencies(projectContents, nodeModules)
+    function exploreDeps(filename: string): void {
+      if (!result.has(filename)) {
+        result.add(filename)
+        const reverseDeps = directReverseDeps[filename]
+        if (reverseDeps != null) {
+          reverseDeps.forEach(exploreDeps)
+        }
+      }
+    }
+    filenamesThatChanged.forEach(exploreDeps)
+    return Array.from(result).sort()
+  } else {
+    return []
+  }
+}
+
+export function identifyFilesThatHaveChanged(
+  oldContents: ProjectContentTreeRoot,
+  newContents: ProjectContentTreeRoot,
+): Set<string> {
+  let filesWithChanges: Set<string> = emptySet()
+
+  function addFileToChanges(filename: string): void {
+    filesWithChanges.add(filename)
+  }
+
+  function applyChanges(
```

Comment:
It looks like this function was extracted from `vscode-bridge.ts` but still exists there? Should `vscode-bridge` now be using this instead?

Reply:
It was duplicated from there but is different internally, it would be possible to make a version that identifies _all_ changes as something like Added/Removed/Modified/Etc and then process those, but I was a bit dubious of abstracting that with only 2 uses.